### PR TITLE
Libraries consistency:

### PR DESCRIPTION
**Pull request by grandizzy**
- refactor `LPs` naming in functions, events and custom error naming to `LP`
- rename `start` reserve auction to `kick` reserve auction. add `KickReserveAuction` event
- calculate `LUP` in single place `Deposits.getLup()`, remove duplicate functions to calculate `LUP`
- split `Auctions` library in `KickerActions`, `TakerActions` and `SettlerActions` libraries
- split `LenderActions` in `LenderActions` and `LPActions` external libs
- cleanup interfaces, restructure them based on actors and actions:
  - `IPoolLPActions` interface split from `IPoolLenderActions`
  - `IPoolLiquidationActions` and `IPoolReserveAuctionActions` interfaces restructured in `IPoolSettlerActions`, `IPoolKickerActions`, `IPoolTakerActions` to match functionality of external libs
- no logic / tests change

deployment cost before change

| Contract | Deployment Cost | Deployment Size |
| --- | --- | --- |
| ERC20Pool | 4798719 | 23893 |
| ERC20PoolFactory | 5385851 | 26716 |
| ERC721Pool | 4924499 | 24521 |
| ERC721PoolFactory | 5685292 | 28215 |

deployment cost after change
| Contract | Deployment Cost | Deployment Size |
| --- | --- | --- |
| ERC20Pool | 4800119 | 23900 |
| ERC20PoolFactory | 5387251 | 26723 |
| ERC721Pool | 4926299 | 24530 |
| ERC721PoolFactory | 5687092 | 28224 |

contract size before change
```
============ Deployment Bytecode Sizes ============
  ERC721Pool               -  24,482B  (99.61%)
  ERC20Pool                -  23,854B  (97.06%)
  Auctions                 -  21,060B  (85.69%)
```
contract size post change
```
============ Deployment Bytecode Sizes ============
  ERC721Pool               -  24,492B  (99.65%)
  ERC20Pool                -  23,862B  (97.09%)
  PositionManager          -  18,539B  (75.43%)
```

TODO: update docs / diagrams

### CODE REVIEW

File: tests/forge/ERC20Pool/ERC20PoolFactory.t.sol [hunk 1]
```diff
@@ -11,7 +11,7 @@ import { IPoolErrors }      from 'src/interfaces/pool/commons/IPoolErrors.sol';
 import { IPoolFactory }     from 'src/interfaces/pool/IPoolFactory.sol';
 
 contract ERC20PoolFactoryTest is ERC20HelperContract {
-    address immutable poolAddress = 0xeCAF6d240E0AdcaD5FfE4306b7D4301Df130bC02;
+    address immutable poolAddress = 0x3Cd0C8d97cD0291E178560E5675dD27AeD0A90d1;
```
Comment (ith-harvey) [inline tests/forge/ERC20Pool/ERC20PoolFactory.t.sol]: My understanding is this change is the result of how the libraries are composed, is that correct?

Comment (grandizzy) [reply tests/forge/ERC20Pool/ERC20PoolFactory.t.sol]: yes, that's correct, the address depends on that

File: src/libraries/external/LPsOwnerActions.sol [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+import { MAX_FENWICK_INDEX } from '../helpers/PoolHelper.sol';
+
+import { Maths } from '../internal/Maths.sol';
+
+/**
+    @title  LPsOwnerActions library
+    @notice External library containing logic for pool actors to:
+            - increase/decrease/revoke LP allowance; approve/remove LP transferors; transfer LPs
+ */
+library LPsOwnerActions {
```
Comment (ith-harvey) [inline src/libraries/external/LPsOwnerActions.sol]: NIT: not crazy about the name `LPs` here... as Owner is singular and LPs stands for liquidity providers. Would suggest `LPOwnerActions`. Same goes for all other references of the `LPs` naming convention. Should be named to `LP`

File: src/libraries/external/LPsOwnerActions.sol [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+     *          - delete allowance mapping
+     *          - increment new lender.lps accumulator and lender.depositTime state
+     *          - delete old lender from bucket -> lender mapping
+     *  @dev reverts on:
+     *          - invalid index InvalidIndex()
+     *          - no allowance NoAllowance()
+     *  @dev emit events:
+     *          - TransferLPs
+     */
+    function transferLPs(
```
Comment (ith-harvey) [inline src/libraries/external/LPsOwnerActions.sol]: I like how `transferLPs` is now included with other LP methods where as before it was off in `lenderActions` by itself.

File: src/libraries/external/KickerActions.sol [hunk 4]
```diff
@@ -0,0 +1,482 @@
...
+     *  @notice See `IPoolReserveAuctionActions` for descriptions.
+     *  @dev    write state:
+     *              - update reserveAuction.unclaimed accumulator
+     *              - update reserveAuction.kicked timestamp state
+     *  @dev    reverts on:
+     *          - no reserves to claim NoReserves()
+     *  @dev    emit events:
+     *              - ReserveAuction
+     */
+    function startClaimableReserveAuction(
```
Comment (ith-harvey) [inline src/libraries/external/KickerActions.sol]: Should this function match the naming convention of kick above?

Possibly rename to `kickReserve`?

File: src/libraries/external/KickerActions.sol [hunk 5]
```diff
@@ -0,0 +1,482 @@
...
+        reserveAuction_.unclaimed = curUnclaimedAuctionReserve;
+        reserveAuction_.kicked    = block.timestamp;
+
+        // increment latest burn event epoch and update burn event timestamp
+        latestBurnEpoch += 1;
+
+        reserveAuction_.latestBurnEventEpoch = latestBurnEpoch;
+        reserveAuction_.burnEvents[latestBurnEpoch].timestamp = block.timestamp;
+
+        emit ReserveAuction(
```
Comment (ith-harvey) [inline src/libraries/external/KickerActions.sol]: In addition to the above I think this event name would be better suited as `kickReserve`

**Review by ith-harvey — APPROVED** [review]
By breaking the code into distinct actor roles it not only matches our previous pattern for pool logic Lender and Borrowers but makes the hardest logic of the code easier to reason about IMO.

File: src/libraries/external/LPsOwnerActions.sol (hunk 2 above)
Comment (grandizzy) [reply src/libraries/external/LPsOwnerActions.sol]: done in https://github.com/ajna-finance/contracts/pull/669/commits/8b8d19260eee0245695fbb282b8cd0713fbec808

File: src/libraries/external/KickerActions.sol (hunk 4 above)
Comment (grandizzy) [reply src/libraries/external/KickerActions.sol]: done in https://github.com/ajna-finance/contracts/pull/669/commits/8b8d19260eee0245695fbb282b8cd0713fbec808

File: src/libraries/external/KickerActions.sol (hunk 5 above)
Comment (grandizzy) [reply src/libraries/external/KickerActions.sol]: done in https://github.com/ajna-finance/contracts/pull/669/commits/8b8d19260eee0245695fbb282b8cd0713fbec808

File: docs/Functions.md [hunk 6]
```diff
@@ -232,25 +232,25 @@
 	emit events:
 	- BondWithdrawn
 
-### startClaimableReserveAuction
+### kickReserveAuction
```
Comment (ith-harvey) [inline docs/Functions.md:235]: :pray:

File: src/PositionManager.sol [hunk 7]
```diff
@@ -197,7 +197,7 @@ contract PositionManager is ERC721, PermitERC721, IPositionManager, Multicall, R
         }
 
         // update pool lps accounting and transfer ownership of lps to PositionManager contract
```
Comment (ith-harvey) [inline src/PositionManager.sol]: NIT: revise comment to -> // update pool LP accounting and transfer ownership of LP to PositionManager contract

File: src/libraries/external/LPOwnerActions.sol [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+import { MAX_FENWICK_INDEX } from '../helpers/PoolHelper.sol';
+
+import { Maths } from '../internal/Maths.sol';
+
+/**
+    @title  LPOwnerActions library
+    @notice External library containing logic for LP owners to:
+            - increase/decrease/revoke LP allowance; approve/revoke LP transferors; transfer LPs
+ */
+library LPOwnerActions {
```
Comment (ith-harvey) [inline src/libraries/external/LPOwnerActions.sol]: NIT: suggest renaming to `LPActions`

**Review by prat-gpt — APPROVED** [review]
lgtm.

**Review by ith-harvey — APPROVED** [review]
Noticed a discrepancy between `kickMomp` and `kickPrice` which is also purely naming related. Don't care which we use as long as its the same throughout the repo.

kickMomp -> https://github.com/ajna-finance/contracts/blob/a65bd98bfdd0c4dc863e6a6827e228abe3d72f80/src/libraries/external/Auctions.sol#L777

kickPrice -> https://github.com/ajna-finance/contracts/blob/65bcd8ea791f4c70452768e4ebe15efd8f1430b6/src/interfaces/pool/commons/IPoolState.sol#L32

If you want to handle that cosmetic change here great, if you think it should be in a separate PR I'm fine with that too.

File: src/libraries/external/LPOwnerActions.sol (hunk 8 above)
Comment (grandizzy) [reply src/libraries/external/LPOwnerActions.sol]: can do that, what would `LP` stand from, is that `LiquidityProvider`?

**Comment by grandizzy** [comment]
> Noticed a discrepancy between `kickMomp` and `kickPrice` which is also purely naming related. Don't care which we use as long as its the same throughout the repo.
> 
> kickMomp ->
> 
> https://github.com/ajna-finance/contracts/blob/a65bd98bfdd0c4dc863e6a6827e228abe3d72f80/src/libraries/external/Auctions.sol#L777
> 
> kickPrice ->
> 
> https://github.com/ajna-finance/contracts/blob/65bcd8ea791f4c70452768e4ebe15efd8f1430b6/src/interfaces/pool/commons/IPoolState.sol#L32
> 
> If you want to handle that cosmetic change here great, if you think it should be in a separate PR I'm fine with that too.

yes, changed with https://github.com/ajna-finance/contracts/pull/669/commits/4a419241b0432f1698a04f1f42dba6ca76b0d648

File: src/libraries/external/LPOwnerActions.sol (hunk 8 above)
Comment (grandizzy) [reply src/libraries/external/LPOwnerActions.sol]: changed with https://github.com/ajna-finance/contracts/pull/669/commits/4a419241b0432f1698a04f1f42dba6ca76b0d648

File: src/PositionManager.sol (hunk 7 above)
Comment (grandizzy) [reply src/PositionManager.sol]: updated in https://github.com/ajna-finance/contracts/pull/669/commits/4a419241b0432f1698a04f1f42dba6ca76b0d648

File: src/libraries/external/LPOwnerActions.sol (hunk 8 above)
Comment (ith-harvey) [reply src/libraries/external/LPOwnerActions.sol]: > 

Yes

**Review by ith-harvey — COMMENTED** [review]
Another cleanup related item... Noticed that the view functions in `Buckets.sol` don't match the `LP` change, additionally in the natspec some of the arg comments need to be aligned.

https://github.com/ajna-finance/contracts/blob/4a419241b0432f1698a04f1f42dba6ca76b0d648/src/libraries/internal/Buckets.sol#L95

**Comment by grandizzy** [comment]
> 

yeah, there are more of that, even some functions with `LPs` in their name but since they're not external I won't go and update all hits in this PR. Could follow up with a different PR

File: src/interfaces/pool/commons/IPoolState.sol [hunk 9]
```diff
@@ -14,7 +14,7 @@ interface IPoolState {
      *  @return bondFactor   The factor used for calculating bond size.
      *  @return bondSize     The bond amount in quote token terms.
      *  @return kickTime     Time the liquidation was initiated.
-     *  @return kickPrice    Highest Price Bucket at time of liquidation.
+     *  @return kickMomp     Highest Price Bucket at time of liquidation (MOMP).
```
Comment (ith-harvey) [inline src/interfaces/pool/commons/IPoolState.sol]: I don't think that this comment accuratley describes this variable. See where kickMomp is set here -> https://github.com/ajna-finance/contracts/blob/develop/src/libraries/external/Auctions.sol#L863-L864

I would suggest we change this comment to:
`The price where the average loan utilizes deposit, at the time when the loan is liquidated (kicked)`

Comment (grandizzy) [reply src/interfaces/pool/commons/IPoolState.sol]: updated in https://github.com/ajna-finance/contracts/pull/669/commits/fb35f1ee6a97563014d0aef15183d29210e47d4d

**Comment by grandizzy** [comment]
> > 
> 
> yeah, there are more of that, even some functions with `LPs` in their name but since they're not external I won't go and update all hits in this PR. Could follow up with a different PR

@ith-harvey updated more `LPs` to `LP` with https://github.com/ajna-finance/contracts/pull/669/commits/18cc6eb8960bea2735f878157f97585e1e9ebbd3 please check

**Comment by ith-harvey** [comment]
> > > 
> 
> > 
> 
> > yeah, there are more of that, even some functions with `LPs` in their name but since they're not external I won't go and update all hits in this PR. Could follow up with a different PR
> 
> 
> 
> @ith-harvey updated more `LPs` to `LP` with https://github.com/ajna-finance/contracts/pull/669/commits/18cc6eb8960bea2735f878157f97585e1e9ebbd3 please check


LGTM ship it

**Review by MikeHathaway — APPROVED** [review]
LGTM! Much easier to navigate the code like this

**Review by ith-harvey — APPROVED** [review]

**Pull request merged**